### DISCUSSION TITLE
(Huge) difference between local build and CI build

### DISCUSSION THREAD

**Discussion by tbroyer** [keycloak/keycloak#12407] [Q&A]
Hi,

I'm building a Docker image (not using the official ones, making my own using a similar process: create user, download distribution from GitHub, validate checksum, untar), using a base image with just the unpacked distribution, and then a more specific image that calls `kc.sh build --db postgres --health-enabled true --metrics-enabled true`.

When I build the Docker image locally (from the base image deployed to our registry), that step creates a 1.1MB layer that only modifies the files in `$KEYCLOAK_HOME/lib/quarkus/`.

However, when the image is built on our CI server, the layer is 192MB ‼ Looking at it with [dive](https://github.com/wagoodman/dive), it tells me many (if not all) JARs in `$KEYCLOAK_HOME/lib/lib/` live in 2 layers, creating waste; this means that the `kc.sh build` has somehow copied them to the new layer, but the "current layer content" view doesn't show them, which means their content is identical and only file system attributes have changed (I haven't found which one would have changed though: `getcap -nrv` doesn't list anything, and `ls -l` lists the same dates, owner, and permissions).

Because the layer was only 1.1MB on my machine, I didn't bother using a multi-stage build like suggested in the docs (and because they're a PITA to correctly cleanup after build, I'd rather keep it that way), but 192MB is a whole different matter!

Anyone knows what `kc.sh build` (i.e. a Quarkus build) does exactly that could explain that?

**Comment by tbroyer** [comment]
So, apparently, this is specific to my machine (Arch Linux, containerd 1.6.5, docker 20.10.16). Coworkers on Ubuntu and Fedora (both containerd 1.6.6, docker 20.10.17) both have the huge layer.

Anyway, I'd love to understand what's happening behind the hood, and even better understand the difference between my machine and others.

**Reply by bs-matil** [reply]
are you using btrfs?

**Reply by tbroyer** [reply]
Nope. We're all on extfs.

Not sure yet if that could be it but I have *Native Overlay Diff: false* while coworkers have it set to `true`:
```
$ docker info
…
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
  userxattr: false
```
